File: 207-course-schedule/course-schedule.cpp
class Solution {
public:
    bool canFinish(int n, vector<vector<int>>& p) {
        
        vector<int>indegree(n,0);
        vector<vector<int>>adj(n);
        for(int i=0;i<p.size();i++){
            int u=p[i][0];
            int v=p[i][1];
            adj[v].push_back(u);
            indegree[u]++;
        }
        queue<int>q;
        for(int i=0;i<n;i++){
            if(indegree[i]==0)q.push(i);
        }
        if(q.empty())return false;
        vector<int>order;
        while(!q.empty()){
            auto node=q.front();
            order.push_back(node);
            q.pop();
            for(auto it:adj[node]){
                indegree[it]--;
                if(indegree[it]==0){
                    q.push(it);
                }
            }

        }
        return n==order.size();


    }
};

File: 207-course-schedule/course-schedule.py
def detect(node,parent,path_vis,vis,adj):
    vis[node]=1
    path_vis[node]=1

    for i in adj[node]:
        if not vis[i]:
            if detect(i,node,path_vis,vis,adj):
                return True
        elif path_vis[i] :
            return True
    
    path_vis[node]=0
    return False




class Solution:
    def canFinish(self, n: int, p: List[List[int]]) -> bool:

        adj=[[] for _ in range(n)]
        vis=[0]*n
        path_vis=[0]*n
        for v,u in p:
            adj[u].append(v)
        
        for i in range(n):
            if not vis[i]:
                if detect(i,-1,path_vis,vis,adj):
                    return False
        
        return True


        